ouroboros-consensus/src/Ouroboros/Consensus/Storage/LedgerDB/HD/LMDB.hs

```diff
@@ -0,0 +1,545 @@
+           -- TODO Should throw a DbErrDbExists exception
+           -- Callers should delete the directory if they want to restore a snapshot
+           FS.removeDirectoryRecursive fs path
+    False | GDDMustExist <- gdd -> Exn.throw $ DbErrDbDoesntExist path
+    _ -> pure ()
+  FS.createDirectoryIfMissing fs True path
+  FS.unsafeToFilePath fs path
+
+{-------------------------------------------------------------------------------
+ Initialize an LMDB
+-------------------------------------------------------------------------------}
+
+-- | How to initialise an LMDB Backing store
+data LMDBInit l
+  = LIInitialiseFromMemory (WithOrigin SlotNo) (LedgerTables l ValuesMK) -- ^ Initialise with these values
+  | LIInitialiseFromLMDB FS.FsPath -- ^ Initialise by copying from an LMDB db at this path
+  | LINoInitialise -- ^ The database is already initialised
+
+-- | Initialize the LMDB from these provided values
+initFromVals ::
+     TableStuff l
+  => WithOrigin SlotNo
+  -> LedgerTables l ValuesMK
+  -> Db l
+  -> IO ()
+initFromVals dbsSeq vals Db{..} = LMDB.readWriteTransaction dbEnv $
+  withDbSettingsRWMaybeNull dbSettings $ \case
+    Nothing -> zipLedgerTablesA initLMDBTable dbBackingTables vals $> ((), DbState{dbsSeq})
+    Just _ -> Exn.throw $ DbErrStr "initFromVals: db already had state"
+
+initFromLMDBs :: MonadIO m
+  => Trace.Tracer IO TraceDb
+  -> FS.SomeHasFS m
+  -> FS.FsPath
+  -> FS.FsPath
+  -> m ()
+initFromLMDBs tracer shfs from0 to0 = do
+  liftIO $ Trace.traceWith tracer $ TDBInitialisingFromLMDB from0
+  from <- guardDbDir GDDMustExist shfs from0
+  to <- guardDbDir GDDMustNotExist shfs to0
+  liftIO $ Exn.bracket
+    (LMDB.openEnvironment from defaultLMDBLimits) -- TODO assess limits, in particular this could fail if the db is too big
+    LMDB.closeEnvironment $ \e -> lmdbCopy tracer e to
+
+
+lmdbCopy :: MonadIO m
+  => Trace.Tracer m TraceDb
+  -> LMDB.Environment LMDB.ReadWrite
+  -> FilePath
+  -> m ()
+lmdbCopy tracer dbEnv to = do
+  -- TODO This copying is gross. Tests depend on it, but I wish I was only responsible for copying the database here
+    let LMDB.Env e = dbEnv
+    from <- liftIO $ LMDB.mdb_env_get_path e
+    Trace.traceWith tracer $ TDBCopying from to
+    liftIO $ LMDB.mdb_env_copy e to
+    -- others <- liftIO $ Dir.listDirectory from <&> filter ((/= ".mdb") . Dir.takeExtension)
+    -- for_ others $ \f0 -> do
+    --   let f = from Dir.</> f0
+    --       t = to Dir.</> f0
+    --   Trace.traceWith tracer $ TDBCopyingFile f t
+    --   liftIO $ Dir.copyFile f t
+    Trace.traceWith tracer $ TDBCopied from to
+
+-- TODO We don't know how much time is spent in I/O
+
+-- TODO 50% of total time is spent somewhere else. Where?
+
+-- | Initialise a backing store
+newLMDBBackingStore :: forall m l. (TableStuff l, MonadIO m, IOLike m)
+  => Trace.Tracer m TraceDb
+  -> LMDBLimits
+  -> FS.SomeHasFS m
+  -> LMDBInit l
+  -> m (LMDBBackingStore l m)
+newLMDBBackingStore tracer limits sfs init_db = do
```

Comment:
We need to consider documenting some parameters. In particular `sfs` could have a more descriptive name.